File: Frontend/web/lib/api/images.ts
import axios from 'axios';

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

export interface ImageResponse {
  id: string;
  listing_id: string;
  filename: string;
  original_url: string;
  thumbnail_url?: string;
  medium_url?: string;
  display_order: number;
  alt_text?: string;
  width?: number;
  height?: number;
  file_size?: number;
  is_main: boolean;
  created_at: string;
}

export interface ImageUploadResponse {
  id: string;
  url: string;
  thumbnail_url?: string;
  message: string;
}

/**
 * Subir una imagen para un listing
 */
export async function uploadImage(listingId: string, file: File, token: string): Promise<ImageUploadResponse> {
  const formData = new FormData();
  formData.append('file', file);

  const response = await axios.post(
    `${API_URL}/v1/images/${listingId}/upload`,
    formData,
    {
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'multipart/form-data',
      },
    }
  );

  return response.data;
}

/**
 * Obtener todas las imágenes de un listing
 */
export async function getListingImages(listingId: string): Promise<ImageResponse[]> {
  const response = await axios.get(`${API_URL}/v1/images/${listingId}`);
  return response.data;
}

/**
 * Actualizar metadatos de una imagen
 */
export async function updateImage(
  imageId: string,
  data: { display_order?: number; alt_text?: string; is_main?: boolean },
  token: string
): Promise<ImageResponse> {
  const response = await axios.put(
    `${API_URL}/v1/images/${imageId}`,
    data,
    {
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    }
  );
  return response.data;
}

/**
 * Eliminar una imagen
 */
export async function deleteImage(imageId: string, token: string): Promise<void> {
  await axios.delete(`${API_URL}/v1/images/${imageId}`, {
    headers: {
      'Authorization': `Bearer ${token}`,
    },
  });
}

/**
 * Marcar una imagen como principal
 */
export async function setMainImage(imageId: string, token: string): Promise<ImageResponse> {
  const response = await axios.post(
    `${API_URL}/v1/images/${imageId}/set-main`,
    {},
    {
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    }
  );
  return response.data;
}

/**
 * Reordenar imágenes de un listing
 */
export async function reorderImages(
  listingId: string,
  imageIds: string[],
  token: string
): Promise<ImageResponse[]> {
  const response = await axios.post(
    `${API_URL}/v1/images/${listingId}/reorder`,
    imageIds,
    {
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    }
  );
  return response.data;
}

File: MAEBACKEND/backend_api/app/api/endpoints/images.py
from fastapi import APIRouter, Depends, HTTPException, status, UploadFile, File
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.schemas.images import ImageResponse, ImageUpdate, ImageUploadResponse
from app.services.image_service import ImageService
from app.api.deps import get_current_user
from typing import List

router = APIRouter()

@router.post("/{listing_id}/upload", response_model=ImageUploadResponse, status_code=status.HTTP_201_CREATED)
async def upload_image(
    listing_id: str,
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """
    Subir una imagen para un listing.
    Formatos permitidos: JPG, JPEG, PNG, WEBP, GIF
    """
    try:
        service = ImageService(db)
        image = service.upload_image(listing_id, file)
        
        return ImageUploadResponse(
            id=str(image.id),
            url=image.original_url,
            thumbnail_url=image.thumbnail_url,
            message="Image uploaded successfully"
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error uploading image: {str(e)}")

@router.get("/{listing_id}", response_model=List[ImageResponse])
async def get_listing_images(
    listing_id: str,
    db: Session = Depends(get_db)
):
    """Obtener todas las imágenes de un listing"""
    service = ImageService(db)
    images = service.get_listing_images(listing_id)
    return [ImageResponse.from_orm(img) for img in images]

@router.put("/{image_id}", response_model=ImageResponse)
async def update_image(
    image_id: str,
    data: ImageUpdate,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Actualizar metadatos de una imagen (orden, alt_text, is_main)"""
    service = ImageService(db)
    image = service.update_image(image_id, data)
    if not image:
        raise HTTPException(status_code=404, detail="Image not found")
    return ImageResponse.from_orm(image)

@router.delete("/{image_id}", status_code=status.HTTP_200_OK)
async def delete_image(
    image_id: str,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Eliminar una imagen"""
    service = ImageService(db)
    success = service.delete_image(image_id)
    if not success:
        raise HTTPException(status_code=404, detail="Image not found")
    return {"message": "Image deleted successfully"}

@router.post("/{image_id}/set-main", response_model=ImageResponse)
async def set_main_image(
    image_id: str,
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Marcar una imagen como principal"""
    service = ImageService(db)
    image = service.set_main_image(image_id)
    if not image:
        raise HTTPException(status_code=404, detail="Image not found")
    return ImageResponse.from_orm(image)

@router.post("/{listing_id}/reorder", response_model=List[ImageResponse])
async def reorder_images(
    listing_id: str,
    image_ids: List[str],
    db: Session = Depends(get_db),
    current_user = Depends(get_current_user)
):
    """Reordenar imágenes de un listing"""
    service = ImageService(db)
    images = service.reorder_images(listing_id, image_ids)
    return [ImageResponse.from_orm(img) for img in images]
